### PR TITLE
Change database folder label format

### PR DESCRIPTION
**Pull request by calmh**
Changes old 64 byte labels to a two byte index into a label table. This handles labels longer than 64 bytes and is a small performance increase overall.

This was why we needed somewhere to keep in memory state per database instance. The database is converted (and renamed) on open so this is backwards compatible, although the resulting database is of course no longer readable by current versions.

I suggest putting this in a v0.13.

### CODE REVIEW

File: lib/db/leveldb_convert.go [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+			globals++
+		case KeyTypeBlock:
+			hash := key[1+64 : 1+64+32]
+			folder, file := oldFromBlockKey(key)
+			newKey := blockKeyInto(nil, hash, dbi.shortFolderID(folder), string(file))
+			if err := to.Put(newKey, i.Value(), nil); err != nil {
+				return err
+			}
+			blocks++
+		default:
```
Comment (AudriusButkevicius) [inline lib/db/leveldb_convert.go]: Needs to convert KeyTypeVirtualMtime, KeyTypeFolderStatistic.

**Comment by AudriusButkevicius** [comment]
So it looks fine, apart from that one comment, but given we've agreed that 1.0 is supposed to have folder labels (as it's a protocol change), perhaps we should not use integers and use folder ID's, in whatever shape or form we device they come in.

**Comment by calmh** [comment]
Hmm, yes, probably. Assuming we want to keep the folder IDs random and reasonably collision free I was going to suggest a random int64.

**Comment by AudriusButkevicius** [comment]
yeah, though it needs to be reproducible random, atleast for the migration part, so that we wouldn't break existing shares.

**Comment by calmh** [comment]
Actually I think I want to push for this PR pretty much as is anyway. My rationale is that it's a bugfix (currently you'll panic or corrupt the db if you create a folder with a label >64 bytes), and that it'll work after the folder label change anyway. In this, we map string folder IDs to uint16s. After folder labels, we may map uint64s to uint16s. It's less of a gain in that case, but it doesn't mean we need to redo the database format. Individual devices should still not exceed 65k folders I think.

**Comment by AudriusButkevicius** [comment]
So the problem I have with this is eventually having 3 different ids for folders.

**Comment by calmh** [comment]
Three?

**Comment by AudriusButkevicius** [comment]
Label, numeric id in index, and N byte id between devices.

**Comment by calmh** [comment]
The label is just a label though, it's just metadata (in the future, when the label exists). Having the database ID be different from the protocol/user-visible ID is no big deal to me. Had this been an SQL database we would certainly have had an ID column on the folders table.

**Comment by AudriusButkevicius** [comment]
So my comment about conversion of virtualmtimes is still valid. Also, 65k folders is our new limit? Otherwise I am happy to merge

**Comment by calmh** [comment]
Sorry, I missed your comment here, will address. About the limit, do you think 65k is an unreasonable limit? We could use an int32 as well I guess, to avoid any risk of ever approaching the limit (although I think tens of thousands of folders is already pretty fucked up...)

**Comment by Zillode** [comment]
+1 for in32

**Comment by calmh** [comment]
I'll use int64 with varint encoding (1-10 bytes, most will be one or two bytes at the most).

**Comment by calmh** [comment]
Actually no, that's stupid and too complex. `uint32` will do just fine.

@AudriusButkevicius The virtual mtime things can't easily be converted, because the key format is `{KeyTypeVirtualMtime, folder, file}` where both folder and file are variable length. That's fine for looking up when you know what you're looking for, but it's not trivial to determine the split between folder and file when just iterating over it. Instead I just drop these entries; there will be a bunch of mtime updates when the update hits Android, I don't think this is catastrophic. I decided not to change the format of the statistics key types as those are like two or three values in total and don't have easy access to the indexing for folders and files. We can expose it if necessary, but I don't think we need to for this purpose.

This passes integration tests for me and looks good, but I may have missed something... It's for v0.13 obviously.

**Comment by calmh** [comment]
If this merges I intend to see if the same thing makes sense for file names; although that index will need to live in the database and not primarily in memory. It should make the block map a lot smaller, as we're currently repeating (possibly quite long) filenames for every block; an uint64 there would be a lot smaller. For loading FileInfos it's one more lookup before getting at the data, and likewise when creating new ones, so it needs some benchmarking.

**Comment by AudriusButkevicius** [comment]
So I am happy with this, should I merge now?

**Comment by calmh** [comment]
Do it! Master becomes v0.13.

**Comment by wweich** [comment]
>Instead I just drop these entries; there will be a bunch of mtime updates when the update hits Android, I don't think this is catastrophic.

Will syncthing try to update mtime on Android then fail and use virtual mtimes again, or will syncthing set the "wrong" mtimes on the other devices?

**Comment by AudriusButkevicius** [comment]
Yes.

**Comment by calmh** [comment]
It's better than just nuking the db and doing a rescan :expressionless:

**Pull request merged**